tests/smoke/amg.cpp

```diff
@@ -0,0 +1,713 @@
+			grb::identities::one >() );
+	(void)rc;
+	assert( rc == SUCCESS );
+
+#ifdef AMG_PRINT_SYSTEM
+	if( spmd<>::pid() == 0 ) {
+		print_vector( x, 50, " ---> X(1)" );
+		print_vector( b, 50, " ---> B(1)" );
+	}
+#endif
+
+	out.times.preamble = timer.time();
+	timer.reset();
+
+	const bool with_preconditioning = ! in.no_preconditioning;
+	out.test_repetitions = 0;
+	if( in.evaluation_run ) {
+		double single_time_start = timer.time();
+		rc = amg( *amg_state, with_preconditioning, in.smoother_steps,
+			in.smoother_steps, in.max_iterations, 0.0, out.performed_iterations, out.residual );
+		double single_time = timer.time() - single_time_start;
+		if( rc == SUCCESS ) {
+			rc = collectives<>::reduce( single_time, 0, operators::max< double >() );
+		}
+		out.times.useful = single_time;
+		out.test_repetitions = static_cast< size_t >( 1000.0 / single_time ) + 1;
+	} else {
+		// do benchmark
+		double time_start = timer.time();
+		for( size_t i = 0; i < in.test_repetitions && rc == SUCCESS; ++i ) {
+			rc = set( x, 0.0 );
+			assert( rc == SUCCESS );
+			rc = amg( *amg_state, with_preconditioning, in.smoother_steps,
+				in.smoother_steps, in.max_iterations, 0.0, out.performed_iterations, out.residual );
+
+			out.test_repetitions++;
+			if( rc != SUCCESS ) {
+				break;
+			}
+
+		}
+		double time_taken = timer.time() - time_start;
+		rc = rc ? rc : collectives<>::reduce( time_taken, 0, operators::max< double >() );
+		out.times.useful = time_taken / static_cast< double >( out.test_repetitions );
+
+#ifdef AMG_PRINT_SYSTEM
+		rc = rc ? rc : grb::eWiseLambda( [ &x ]( const size_t i ) { x[ i ] -= 1;}, x );
+		print_norm (x, " norm(x)");
+#endif
+		// sleep( 1 );
+	}
+	timer.reset();
+
+#ifdef AMG_PRINT_SYSTEM
+	if( spmd<>::pid() == 0 ) {
+		print_vector( x, 50, " x(first 50 elements)" );
+		print_vector( b, 50, " b(first 50 elements)" );
+	}
+#endif
+
+	if( spmd<>::pid() == 0 ) {
+		if( rc == SUCCESS ) {
+			if( in.evaluation_run ) {
+				std::cout << "Info: cold AMG completed within " << out.performed_iterations
+				          << " iterations. Last computed residual is " << out.residual
+				          << ". Time taken was " << out.times.useful
+				          << " ms. Deduced inner repetitions parameter of " << out.test_repetitions
+				          << " to take 1 second or more per inner benchmark." << std::endl;
+			} else {
+				std::cout << "Final residual= "<< out.residual << " relative error= "
+				          <<  out.residual/sqrt(norm_b) << "\n";
+				std::cout << "Average time taken for each of " << out.test_repetitions
+				          << " AMG calls (hot start): " << out.times.useful << std::endl;
+			}
+		} else {
+			std::cerr << "Failure: call to AMG did not succeed (" << toString( rc ) << ")." << std::endl;
+		}
+	}
+
+	// start postamble
+	timer.reset();
+	// set error code
+	out.error_code = rc;
+
+	Semiring< grb::operators::add< double >,
+		grb::operators::mul< double >,
+		grb::identities::zero,
+		grb::identities::one > ring;
```

Comment:
here's that ring! hahah. Move upwards I guess, then reuse:). Also add enter before `>`